### ISSUE TITLE
Consumers fail to commit offsets after rebalancing

### ISSUE THREAD

**Issue by phoebesimon** [SOHU-Co/kafka-node#122]
Hi all,

I'm running into a problem with the highLevelConsumer after rebalancing. Here's the scenario:
- Create two consumers for a topic
- Use a producer to push messages to the topic
- Consumers consume messages fine
- Add a third consumer for the topic
- Rebalance attempt is successful
- Push more messages to the topic
- At this point, I've noticed that some of the consumers did not commit their offsets from the first round of messages, so the rebalanced consumers end up consuming messages again

I guess my first question is: am I doing this correctly and should that work the way I expect?

This doesn't happen every time. I suspect there's a race condition in the `rebalanceAttempt` function. I think it's possible for this line (https://github.com/SOHU-Co/kafka-node/blob/master/lib/highLevelConsumer.js#L274) to be run before the other consumers have finished committing their offsets.

Let me know if I can provide anything else to help.

**Comment by MaciekRakowski** [comment]
If I'm understanding this issue correctly, you explicitly called 'commit', the cluster rebalances, and the offset is not committed. I was also experiencing this issue. After further digging, and comparing with the implementation of the Java high level consumer, it seems that one key difference is that standard Java high level consumer commits synchronously whereas the NodeJS one commits asynchronously. In my experience, when I stopped consuming, and called "commit", it worked just fine.

That being said, I challenge you to try this:
1. Turn off "auto-commit", and commit manually.
2. When calling "commit" (with the 'force' parameter set to true), stop consuming (call 'pause' on the consumer).
3. When the asynchronous commit call finishes, resume consuming.

In other words, stop consuming messages while a commit is taking place. Do this by either making the commit call synchronous, or pausing the consumer while it's in the middle of a commit. If you try this, and it doesn't work, let me know, and I can either troubleshoot the code, or I can send in a sample consumer application.

**Comment by hyperlink** [comment]
Is this still a problem?

**Comment by fhongtao** [comment]
11
